filebeat/input/filestream/prospector.go

```diff
@@ -124,9 +124,14 @@ func (p *fileProspector) Run(ctx input.Context, s loginp.StateMetadataUpdater, h
 						break
 					}
 				}
-
 				hg.Start(ctx, src)
 
+			case loginp.OpTruncate:
+				log.Debugf("File %s has been truncated", fe.NewPath)
+
+				s.UpdateCursor(src, state{Offset: 0})
+				hg.Restart(ctx, src)
```

Comment:
UpdateCursor and Restart one after another potentially add another race condition. The sequence of the race goes like this:

1. harvester publishes event A, offset X
2. updateOp for event A with resource.version 0 is created
3. UpdateCursor increases resource version to 1
4. harvester publishes event B, offset y
5. updateOp for event B with resource.version 1 is created
6. harvester is stopped and new one is started
7. event A is acked and updateOp gets discarded
8. event B is acked and overwrites the cursor with offset Y


Possible mitigations:
1. Never call UpdateCursor if an harvester is still active.
2. Fix the resource version when taking the resource lock.

e.g.

```
type resource struct {
  version, lockedVersion uint
  ...
}

func lockResource(...) error {
  ...
  r.lockVersion = r.version
} 
```

When we create an `updateOp`, we always must use `lockVersion`. When checking if an updateOp must be discarded, we check against `version`. This way we basically discard all in flight and future cursor updates from the harvester that is still running. The `version` more or less acts as a 'cheap mans' logical clock ignoring everything that is supposed to be older than now. The start `timestamp` for all updateOps emitted by a harvester is fixed to the point in time when the harvester was created. 

Using the versioning approach we in theory don't even need to wait for the old harvester to stop (some more minor refactoring would be required though) when we start the new one. The old harvester would just not be able to update the state stored in the ephemeral and persistent stores.